group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -144,216 +133,137 @@ protected GroupAssignment buildAssignment() throws PartitionAssignorException {
-            );
-
-            int currentAssignmentSize = validCurrentMemberAssignment.size();
-            // Number of partitions required to meet the minimum quota.
-            int remaining = minQuota - currentAssignmentSize;
-
-            if (currentAssignmentSize > 0) {
-                int retainedPartitionsCount = min(currentAssignmentSize, minQuota);
-                IntStream.range(0, retainedPartitionsCount).forEach(i -> {
-                    TopicIdPartition topicIdPartition = validCurrentMemberAssignment.get(i);
-                    addPartitionToAssignment(
-                        targetAssignment,
-                        memberId,
-                        topicIdPartition.topicId(),
-                        topicIdPartition.partitionId()
-                    );
-                });
-
-                if (remaining < 0) {
-                    // The extra partition is located at the last index from the previous step.
-                    if (remainingMembersToGetAnExtraPartition > 0) {
-                        TopicIdPartition topicIdPartition = validCurrentMemberAssignment.get(retainedPartitionsCount++);
-                        addPartitionToAssignment(
-                            targetAssignment,
-                            memberId,
-                            topicIdPartition.topicId(),
-                            topicIdPartition.partitionId()
-                        );
-                        remainingMembersToGetAnExtraPartition--;
+    private void maybeRevokePartitions() {
+        for (Map.Entry<String, AssignmentMemberSpec> entry : groupSpec.members().entrySet()) {
+            String memberId = entry.getKey();
+            AssignmentMemberSpec assignmentMemberSpec = entry.getValue();
+            Map<Uuid, Set<Integer>> oldAssignment = assignmentMemberSpec.assignedPartitions();
+            Map<Uuid, Set<Integer>> newAssignment = null;
+
+            // The assignor expects to receive the assignment as an immutable map. It leverages
+            // this knowledge in order to avoid having to copy all assignments.
+            if (!isImmutableMap(oldAssignment)) {
+                throw new IllegalStateException("The assignor expect an immutable map.");
+            }
+
+            int quota = minimumMemberQuota;
+            if (remainingMembersToGetAnExtraPartition > 0) {
+                quota++;
+                remainingMembersToGetAnExtraPartition--;
+            }
+
+            for (Map.Entry<Uuid, Set<Integer>> topicPartitions : oldAssignment.entrySet()) {
+                Uuid topicId = topicPartitions.getKey();
+                Set<Integer> partitions = topicPartitions.getValue();
+
+                if (subscribedTopicIds.contains(topicId)) {
+                    if (partitions.size() <= quota) {
+                        quota -= partitions.size();
+                    } else {
+                        for (Integer partition : partitions) {
```

Comment:
does this cause randomness? i'm wondering since we iterate over a set we will revoke different partitions each time. and does this matter?

Reply:
Nop it doesn't really matter what you revoke, esp not without rack awareness enabled

Reply:
yep It might make the results unpredictable, in which case we probably can't/shouldn't do 1:1 mapping in the tests

Reply:
Do we depend on LInkedHashSet for ordering in tests?

Reply:
We do depend on LInkedHashSet in the tests in order to make the output predictable. However, the order does not matter at all here. The only property that we must guarantee is that we revoke some partitions if the quota is exceeded.